File: bitgn_run/artifacts.py
from __future__ import annotations

import json
from pathlib import Path
from typing import Any

from .config import RunConfig
from .types import TaskResult


class ArtifactWriter:
    def __init__(self, config: RunConfig) -> None:
        self.run_dir = config.artifact_dir / config.run_id
        self.run_dir.mkdir(parents=True, exist_ok=True)
        write_json(
            self.run_dir / "run_config.json",
            {
                "run_id": config.run_id,
                "run_name": config.run_name,
                "env": config.env,
                "version": config.version,
                "leaderboard": config.leaderboard,
                "fail_fast": config.fail_fast,
                "workers": config.workers,
                "tasks": config.tasks,
                "enabled_rules": sorted(config.enabled_rules),
                "max_wall_sum_seconds": config.max_wall_sum_seconds,
            },
        )

    def append_task(self, result: TaskResult) -> None:
        append_jsonl(self.run_dir / "run_manifest.jsonl", result.to_json())

    def finish(self, results: list[TaskResult]) -> None:
        passed = sum(1 for result in results if result.passed)
        wall_sum = sum(result.wall_seconds or 0.0 for result in results)
        write_json(
            self.run_dir / "run_summary.json",
            {
                "tasks_total": len(results),
                "passed": passed,
                "failed": len(results) - passed,
                "pass_rate": passed / len(results) if results else 0.0,
                "task_wall_time_sum_seconds": wall_sum,
                "results": [result.to_json() for result in results],
            },
        )


def write_json(path: Path, payload: dict[str, Any]) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_text(json.dumps(payload, ensure_ascii=True, indent=2, default=str) + "\n", encoding="utf-8")


def append_jsonl(path: Path, payload: dict[str, Any]) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    with path.open("a", encoding="utf-8") as stream:
        stream.write(json.dumps(payload, ensure_ascii=True, default=str) + "\n")

File: src/artifacts.rs
use crate::config::RunConfig;
use crate::types::TaskResult;
use serde_json::json;
use std::fs::{self, OpenOptions};
use std::io::Write;
use std::path::{Path, PathBuf};

pub struct ArtifactWriter {
    run_dir: PathBuf,
}

impl ArtifactWriter {
    pub fn new(config: &RunConfig) -> Result<Self, String> {
        let run_dir = config.artifact_dir.join(&config.run_id);
        fs::create_dir_all(&run_dir).map_err(|e| e.to_string())?;
        let payload = json!({
            "run_id": config.run_id,
            "run_name": config.run_name,
            "env": config.env,
            "version": config.version,
            "leaderboard": config.leaderboard,
            "fail_fast": config.fail_fast,
            "workers": config.workers,
            "tasks": config.tasks,
            "enabled_rules": config.enabled_rules,
            "max_wall_sum_seconds": config.max_wall_sum_seconds,
        });
        write_json(&run_dir.join("run_config.json"), &payload)?;
        Ok(Self { run_dir })
    }

    pub fn append_task(&self, result: &TaskResult) -> Result<(), String> {
        append_jsonl(&self.run_dir.join("run_manifest.jsonl"), result)
    }

    pub fn finish(&self, results: &[TaskResult]) -> Result<(), String> {
        let passed = results.iter().filter(|r| r.passed == Some(true)).count();
        let failed = results
            .iter()
            .filter(|r| r.score_available && r.passed != Some(true))
            .count();
        let unscored = results.iter().filter(|r| !r.score_available).count();
        let wall_sum: f64 = results.iter().filter_map(|r| r.wall_seconds).sum();
        let payload = json!({
            "tasks_total": results.len(),
            "passed": passed,
            "failed": failed,
            "unscored": unscored,
            "pass_rate": if results.is_empty() { 0.0 } else { passed as f64 / results.len() as f64 },
            "task_wall_time_sum_seconds": wall_sum,
            "results": results,
        });
        write_json(&self.run_dir.join("run_summary.json"), &payload)
    }
}

pub fn write_json(path: &Path, payload: &serde_json::Value) -> Result<(), String> {
    let text = serde_json::to_string_pretty(payload).map_err(|e| e.to_string())?;
    fs::write(path, format!("{text}\n")).map_err(|e| e.to_string())
}

fn append_jsonl<T: serde::Serialize>(path: &Path, payload: &T) -> Result<(), String> {
    let mut file = OpenOptions::new()
        .create(true)
        .append(true)
        .open(path)
        .map_err(|e| e.to_string())?;
    let line = serde_json::to_string(payload).map_err(|e| e.to_string())?;
    writeln!(file, "{line}").map_err(|e| e.to_string())
}
